persist/fs/files.go

```diff
@@ -83,17 +88,52 @@ func (a byTimeAscending) Less(i, j int) bool {
 	return ti.Before(tj)
 }
 
+// byTimeAndIndexAscending sorts files by their block start times and index in ascending
+// order. If the files do not have block start times or indexes in their names, the result
+// is undefined.
+type byTimeAndIndexAscending []string
+
+func (a byTimeAndIndexAscending) Len() int      { return len(a) }
+func (a byTimeAndIndexAscending) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a byTimeAndIndexAscending) Less(i, j int) bool {
+	ti, ii, _ := TimeAndIndexFromFileName(a[i])
+	tj, ij, _ := TimeAndIndexFromFileName(a[j])
+	if ti.Before(tj) {
+		return true
+	}
+	return ti.Equal(tj) && ii < ij
+}
+
 // TimeFromFileName extracts the block start time from file name.
 func TimeFromFileName(fname string) (time.Time, error) {
 	components := strings.Split(filepath.Base(fname), separator)
 	if len(components) < 2 {
 		return timeZero, fmt.Errorf("unexpected file name %s", fname)
 	}
-	latestTimeInNano, err := strconv.ParseInt(components[0], 10, 64)
+	str := strings.Replace(components[1], fileSuffix, "", 1)
+	nanoseconds, err := strconv.ParseInt(str, 10, 64)
 	if err != nil {
 		return timeZero, err
 	}
-	return time.Unix(0, latestTimeInNano), nil
+	return time.Unix(0, nanoseconds), nil
+}
+
+// TimeAndIndexFromFileName extracts the block start and index from file name.
+func TimeAndIndexFromFileName(fname string) (time.Time, int, error) {
+	t, err := TimeFromFileName(fname)
```

Comment:
nit: probably more efficient to parse the time inline below instead of splitting the name twice